java/arcs/core/data/expression/InferredType.kt

```diff
@@ -0,0 +1,123 @@
+package arcs.core.data.expression
+
+import arcs.core.util.BigInt
+import kotlin.reflect.KClass
+
+/**
+ * Matches [widenAndApply] in Expression.kt
+ *
+ * The types are listed in terms of decreasing width, so that a type at position n can always
+ * be widened to a type at position n - 1.
+ */
+private val widenOrder = listOf<InferredType.Primitive>(
+  InferredType.Primitive.NumberType,
+  InferredType.Primitive.DoubleType,
+  InferredType.Primitive.FloatType,
+  InferredType.Primitive.BigIntType,
+  InferredType.Primitive.LongType,
+  InferredType.Primitive.IntType,
+  InferredType.Primitive.ShortType,
+  InferredType.Primitive.ByteType
+)
+
+/** Convert a [KClass] into an appropriate [InferredType] */
+fun KClass<*>.toType() = InferredType.Primitive.of(this)
+
+/** Widen a list of types to the largest type that encompasses all of them. */
+fun widen(vararg types: InferredType): InferredType.Primitive {
+  types.forEach {
+    require(it is InferredType.Numeric) {
+      "$it is not numeric"
+    }
+  }
+
+  for (type in widenOrder) {
```

Comment:
I think `first` with a predicate would be more appropriate here. Although, you'd have to do some wrangling to still throw the illegal arg exception.

Reply:
Good suggestion, I think `firstOrNull() ?: throw ... ` will allow the exception to be handled.